internal/controllers/clusterclass/clusterclass_controller.go

```diff
@@ -177,25 +195,65 @@ func (r *Reconciler) reconcile(ctx context.Context, clusterClass *clusterv1.Clus
 		return ctrl.Result{}, kerrors.NewAggregate(errs)
 	}
 
-	// Ensure the variables are added to the ClusterClass status.
-	clusterClass.Status.Variables = []clusterv1.ClusterClassStatusVariable{}
-	for _, variable := range clusterClass.Spec.Variables {
-		clusterClass.Status.Variables = append(clusterClass.Status.Variables,
-			clusterv1.ClusterClassStatusVariable{
-				Name: variable.Name,
-				Definitions: []clusterv1.ClusterClassStatusVariableDefinition{
-					{
-						From:     clusterv1.VariableDefinitionFromInline,
-						Required: variable.Required,
-						Schema:   variable.Schema,
-					},
-				}})
+	// Reconcile variables
+	if err := r.reconcileVariables(ctx, clusterClass); err != nil {
+		return ctrl.Result{}, err
 	}
+
 	reconcileConditions(clusterClass, outdatedRefs)
 
 	return ctrl.Result{}, nil
 }
 
+func (r *Reconciler) reconcileVariables(ctx context.Context, clusterClass *clusterv1.ClusterClass) error {
+	errs := []error{}
+	uniqueVars := map[string]bool{}
+
+	// Add inline variable definitions to the ClusterClass status.
+	clusterClass.Status.Variables = []clusterv1.ClusterClassStatusVariable{}
+	for _, variable := range clusterClass.Spec.Variables {
+		uniqueVars[variable.Name] = true
+		clusterClass.Status.Variables = append(clusterClass.Status.Variables, statusVariableFromClusterClassVariable(variable, clusterv1.VariableDefinitionFromInline))
+	}
+
+	// If RuntimeSDK is enabled call the DiscoverVariables hook for all associated Runtime Extensions and add the variables
+	// to the ClusterClass status.
+	if feature.Gates.Enabled(feature.RuntimeSDK) {
+		for _, patch := range clusterClass.Spec.Patches {
+			if patch.External == nil || patch.External.DiscoverVariablesExtension == nil {
+				continue
+			}
+			req := &runtimehooksv1.DiscoverVariablesRequest{}
+			req.Settings = patch.External.Settings
+
+			resp := &runtimehooksv1.DiscoverVariablesResponse{}
+			err := r.RuntimeClient.CallExtension(ctx, runtimehooksv1.DiscoverVariables, clusterClass, *patch.External.DiscoverVariablesExtension, req, resp)
+			if err != nil {
+				errs = append(errs, errors.Wrapf(err, "failed to call DiscoverVariables for patch %s", patch.Name))
+			}
+			if resp.Status != runtimehooksv1.ResponseStatusSuccess {
+				errs = append(errs, errors.Errorf("patch %s returned status %q with message %q", patch.Name, resp.Status, resp.Message))
+			}
+			if len(errs) > 0 {
+				continue
+			}
+			if resp.Variables != nil {
+				for _, variable := range resp.Variables {
+					// TODO: Variables should be validated and deduplicated based on their provided definition.
+					if _, ok := uniqueVars[variable.Name]; !ok {
+						errs = append(errs, errors.Errorf("duplicate variable name %s discovered in patch: %s", variable.Name, patch.Name))
```

Comment:
```suggestion
						errs = append(errs, errors.Errorf("duplicate variable name %s discovered in patch: %s", variable.Name, patch.Name))
						continue
```

I would also continue here, just feels cleaner to not continue if we know that something is wrong and not add the variable anyway.

Especially as we are patching the ClusterClass regardless of if we return an error here or not (If I see correctly)

Reply:
Yeah - this is completely a placeholder so it doesn't really matter what happens here for now IMO.

Reply:
I get your point, just trying to have the code on main as clean as we can with the partial state.

Just to avoid that we have something that we forgot to change later and then we end up with a bug